### PR TITLE
add support for Arrow IPC stream format

### PR DESCRIPTION
**Pull request by nwt**
Features:

* Auto-detection via zio/anyio.NewReader*
* Format specification via "-f arrows" and "-i arrows" (".arrows" being the file extension recommend by the Arrow project for this format)
* HTTP API support in requests and responses via the application/vnd.apache.arrow.stream MIME type
* Preservation on read of most Arrow types (excluding metadata) via Zed named types (with an "arrow_" prefix) for Arrow types without an equivalent Zed type
* Ability to write most Arrow types given properly named Zed types (with an "arrow_" prefix)

Limitations:

* No support for the Arrow IPC file format
* Reader is slow due to repeated array.New*Data calls
* Writer does not support big endianness, compression, control of record batch length, or dictionary or sparce union types
* Tests aren't thorough

For #4226.

### CODE REVIEW

**Comment by nwt** [comment]
This increases the zq executable's size by 17 MB for darwin/amd64 and 12 MB for linux/amd64. Ouch.
```console
$ git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
$ git rev-parse @
a31161a3404c7cb45ba2dd405b195c01e0123cf9
$ GOOS=darwin GOARCH=amd64 make build
$ ls -lh dist/zq
-rwxr-xr-x  1 noah  staff    18M Dec  8 07:34 dist/zq*
$ GOOS=linux GOARCH=amd64 make build
$ ls -lh dist/zq
-rwxr-xr-x  1 noah  staff    13M Dec  8 07:34 dist/zq*
$ git checkout arrow
Switched to branch 'arrow'
$ git rev-parse @
2f046e6473c9a85ffa5b4ce19daf8b11a634c364
$ GOOS=darwin GOARCH=amd64 make build
$ ls -lh dist/zq
-rwxr-xr-x  1 noah  staff    35M Dec  8 07:35 dist/zq*
$ GOOS=linux GOARCH=amd64 make build
$ ls -lh dist/zq
-rwxr-xr-x  1 noah  staff    25M Dec  8 07:35 dist/zq*
```

File: zio/arrowio/reader.go [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+		r.rec = nil
+	}
+	return &r.val, nil
+}
+
+var dayTimeIntervalFields = []zed.Column{
+	{Name: "days", Type: zed.TypeInt32},
+	{Name: "milliseconds", Type: zed.TypeUint32},
+}
+var decimal128Fields = []zed.Column{
```
Comment (mccanne) [inline zio/arrowio/reader.go:120]: Let's make an issue for adding decimal128 support.  The type code exists in the spec.

Comment (philrz) [reply zio/arrowio/reader.go:120]: @mccanne: There's an existing issue #1522 that speaks to implementing all four of the decimal type variations. We could open separate issues for each if you'd like, or just keep that one.

File: zio/arrowio/reader.go [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+	case arrow.UINT32:
+		return zed.TypeUint32, nil
+	case arrow.INT32:
+		return zed.TypeInt32, nil
+	case arrow.UINT64:
+		return zed.TypeUint64, nil
+	case arrow.INT64:
+		return zed.TypeInt64, nil
+	case arrow.FLOAT16:
+		return r.zctx.LookupTypeNamed("arrow_float16", zed.TypeFloat32)
```
[the remark this answers is not in the document]
Comment (philrz) [reply zio/arrowio/reader.go:154]: Already open as #1312.

File: zio/arrowio/writer.go [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+
+var (
+	ErrMultipleTypes   = errors.New("arrowio: encountered multiple types (consider 'fuse')")
+	ErrNotRecord       = errors.New("arrowio: not a record")
+	ErrUnsupportedType = errors.New("arrowio: unsupported type")
+)
+
+// Writer is a zio.Writer for the Arrow IPC stream format.  Given Zed values
+// with appropriately named types (see the newArrowDataType implementation), it
+// can write all Arrow types except dictionaries and sparse unions.
```
[the remark this answers is not in the document]
Comment (nwt) [reply zio/arrowio/writer.go]: Done in 868965f161e5633f376c351807bce9f7f2449d23.

**Review by mccanne — APPROVED** [review]
Nice work!

**Pull request merged**